File: src/main/java/com/aaronhowser1/pitchperfect/InstrumentItem.java
package com.aaronhowser1.pitchperfect;

import net.minecraft.core.particles.ParticleTypes;
import net.minecraft.sounds.SoundEvent;
import net.minecraft.sounds.SoundSource;
import net.minecraft.util.Mth;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResultHolder;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;

public class InstrumentItem extends Item {

    public final SoundEvent sound;

    public InstrumentItem(SoundEvent s) {
        super(new Item.Properties()
                .stacksTo(1)
                .tab(ModCreativeModeTab.MOD_TAB)
        );
        sound = s;
    }

    @Override
    public InteractionResultHolder<ItemStack> use(Level level, Player player, InteractionHand interactionHand) {
        ItemStack itemStack = player.getItemInHand(interactionHand);
        float pitch = player.getXRot();
        pitch = map(pitch, -90,90,2,0.5F); //from [-90,90] to [2,0.5], high->low bc big number = low pitch
        playSound(level, pitch, player.getX(), player.getY(), player.getZ());
        spawnNote(level, pitch, player.getX(), player.getY(), player.getZ());
        return InteractionResultHolder.fail(itemStack);
    }

    @Override
    public boolean onLeftClickEntity(ItemStack stack, Player player, Entity entity) {
        float randomPitch = (int) (Math.random() * 180) - 90; //random number [0,180] -> [-90,90]
        randomPitch = map(randomPitch, -90,90,2,0.5F); //from [-90,90] to [2,0.5], high->low bc big number = low pitch
        playSound(entity.getLevel(), randomPitch, entity.getX(), entity.getY(), entity.getZ());
        int randomAmount = (int) Math.floor(Math.random()*8)+2;// [3,10]
        for (int note = 1; note <= randomAmount; note++) {
            double entityWidth = entity.getBbWidth();
            double entityHeight = entity.getBbHeight();
            double noteX = entity.getX()+entityWidth*(Math.random()*3-1.5);
            double noteZ = entity.getZ()+entityWidth*(Math.random()*3-1.5);
            double noteY = entity.getY()+entityHeight+(entityHeight*Math.random()*1.5-.75);
            spawnNote(
                    entity.getLevel(),
                    randomPitch,
                    noteX, noteY, noteZ
            );
        }
        return super.onLeftClickEntity(stack, player, entity);
    }


    public static float map(float value, float min1, float max1, float min2, float max2)
    {
        return min2 + (max2 - min2) * ((value - min1) / (max1 - min1));
    }

    private void playSound(Level level, float pitch, double x, double y, double z) {
        level.playLocalSound(
                x,
                y,
                z,
                sound,
                SoundSource.PLAYERS,
                3, //Volume
                pitch,
                false
        );
    }

    private void spawnNote(Level level, float pitch, double x, double y, double z) {
        float noteColor = map(pitch, 2, 0.5F, 0, 0.5F) + 0.75F;
        level.addParticle(
                ParticleTypes.NOTE,
//                entity.getX(),
//                entity.getEyeY()+entity.getEyeHeight()*0.25,
//                entity.getZ(),
                x, y, z,
                getColor(noteColor, "red"),
                getColor(noteColor, "green"),
                getColor(noteColor, "blue")
        );
        if (false) //set true to debug
            System.out.println("Spawning particle:\nPosition: "+
                    x+","+y+","+z+
                    "\nPitch: "+String.valueOf(pitch)+
                    "\nColor: "+
                    String.valueOf(getColor(noteColor, "red"))+","+
                    String.valueOf(getColor(noteColor, "green"))+","+
                    String.valueOf(getColor(noteColor, "blue"))
            );
    }
    private float getColor(float pitch, String color) {
        return switch (color) {
            case "red" ->
                    Math.max(0.0F, Mth.sin((pitch + 0.0F) * ((float) Math.PI * 2F)) * 0.65F + 0.35F);
            case "blue" ->
                    Math.max(0.0F, Mth.sin((pitch + 0.6666667F) * ((float) Math.PI * 2F)) * 0.65F + 0.35F);
            case "green" ->
                    Math.max(0.0F, Mth.sin((pitch + 0.33333334F) * ((float) Math.PI * 2F)) * 0.65F + 0.35F);
            default -> 0;
        };
    }
}

File: src/main/kotlin/dev/aaronhowser/mods/pitchperfect/item/InstrumentItem.kt
package dev.aaronhowser.mods.pitchperfect.item

import dev.aaronhowser.mods.pitchperfect.advancement.AdvancementTriggers
import dev.aaronhowser.mods.pitchperfect.config.CommonConfig
import dev.aaronhowser.mods.pitchperfect.enchantment.BwaaapEnchantment
import dev.aaronhowser.mods.pitchperfect.enchantment.HealingBeatEnchantment
import dev.aaronhowser.mods.pitchperfect.enchantment.ModEnchantments
import dev.aaronhowser.mods.pitchperfect.item.component.SoundEventComponent
import dev.aaronhowser.mods.pitchperfect.packet.ModPacketHandler
import dev.aaronhowser.mods.pitchperfect.packet.server_to_client.SpawnNotePacket
import dev.aaronhowser.mods.pitchperfect.registry.ModSounds
import dev.aaronhowser.mods.pitchperfect.song.parts.Note
import dev.aaronhowser.mods.pitchperfect.util.OtherUtil.map
import net.minecraft.client.player.LocalPlayer
import net.minecraft.core.Holder
import net.minecraft.network.chat.Component
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.sounds.SoundEvent
import net.minecraft.world.InteractionHand
import net.minecraft.world.InteractionResultHolder
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.EquipmentSlotGroup
import net.minecraft.world.entity.ai.attributes.AttributeModifier
import net.minecraft.world.entity.ai.attributes.Attributes
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.Item
import net.minecraft.world.item.ItemStack
import net.minecraft.world.item.component.ItemAttributeModifiers
import net.minecraft.world.level.Level
import net.minecraft.world.level.block.state.properties.NoteBlockInstrument
import net.minecraft.world.phys.Vec3
import kotlin.random.Random

class InstrumentItem(
    val instrument: SoundEvent
) : Item(
    Properties()
        .durability(100)
        .component(SoundEventComponent.component, SoundEventComponent(instrument))
        .attributes(
            ItemAttributeModifiers.builder()
                .add(
                    Attributes.ATTACK_DAMAGE,
                    AttributeModifier(
                        BASE_ATTACK_DAMAGE_ID,
                        0.1,
                        AttributeModifier.Operation.ADD_VALUE
                    ),
                    EquipmentSlotGroup.MAINHAND
                )
                .build()
        )
) {

    constructor(noteBlockInstrument: NoteBlockInstrument) : this(noteBlockInstrument.soundEvent.value())
    constructor(holder: Holder<SoundEvent>) : this(holder.value())

    companion object {

        fun getSoundEvent(itemStack: ItemStack): SoundEvent? {
            return itemStack.get(SoundEventComponent.component)?.soundEvent
        }

        fun healingBeat(itemStack: ItemStack, player: Player) {
            if (player.cooldowns.isOnCooldown(itemStack.item)) return

            val healingBeatLevel = itemStack.getEnchantmentLevel(
                ModEnchantments.getEnchantHolder(player.level(), ModEnchantments.healingBeatResourceKey)
            )

            if (healingBeatLevel == 0) return

            HealingBeatEnchantment.trigger(player, itemStack)
        }

        fun bwaaap(itemStack: ItemStack, player: Player) {
            if (player.cooldowns.isOnCooldown(itemStack.item)) return

            val bwaaapLevel = itemStack.getEnchantmentLevel(
                ModEnchantments.getEnchantHolder(player.level(), ModEnchantments.bwaaapResourceKey)
            )

            if (bwaaapLevel == 0) return

            BwaaapEnchantment.trigger(player, itemStack)
        }

    }

    override fun getBreakingSound(): SoundEvent = ModSounds.GUITAR_SMASH.get()

    override fun use(
        pLevel: Level,
        pPlayer: Player,
        pUsedHand: InteractionHand
    ): InteractionResultHolder<ItemStack> {
        val itemStack = pPlayer.getItemInHand(pUsedHand)

        useInstrument(itemStack, pPlayer, pLevel, pUsedHand)

        return InteractionResultHolder.pass(itemStack)
    }

    private fun useInstrument(
        itemStack: ItemStack,
        player: Player,
        level: Level,
        interactionHand: InteractionHand
    ) {
        val sound = getSoundEvent(itemStack) ?: return

        val lookVector = player.lookAngle
        val pitch = lookVector.y.toFloat().map(-1f, 1f, 0.5f, 2f)

        val noteVector = if (interactionHand == InteractionHand.MAIN_HAND) {
            lookVector.yRot(-0.5f)
        } else {
            lookVector.yRot(0.5f)
        }

        val bwaaapLevel = itemStack.getEnchantmentLevel(
            ModEnchantments.getEnchantHolder(player.level(), ModEnchantments.bwaaapResourceKey)
        )

        if (!level.isClientSide) {
            ModPacketHandler.messageNearbyPlayers(
                SpawnNotePacket(
                    sound.location,
                    pitch,
                    (player.x + noteVector.x),
                    (player.eyeY + noteVector.y),
                    (player.z + noteVector.z),
                    bwaaapLevel != 0
                ),
                level as ServerLevel,
                player.eyePosition,
                128.0
            )
        }

        healingBeat(itemStack, player)
        bwaaap(itemStack, player)
    }

    override fun onLeftClickEntity(stack: ItemStack, player: Player, entity: Entity): Boolean {
        if (entity.level().isClientSide) return false

        val sound = getSoundEvent(stack) ?: return false

        val particleAmountLowerBound = CommonConfig.MIN_ATTACK_PARTICLES.get()
        val particleAmountUpperBound = CommonConfig.MAX_ATTACK_PARTICLES.get()

        require(particleAmountLowerBound <= particleAmountUpperBound) {
            "Min attack particles cannot be greater than max attack particles."
        }

        val range = particleAmountLowerBound..particleAmountUpperBound
        val randomAmount = range.random()

        val entityWidth = entity.bbWidth
        val entityHeight = entity.bbHeight

        for (note in 1..randomAmount) {
            val randomPitch = Random.nextDouble(0.5, 2.0).toFloat()

            val noteX = entity.x + entityWidth * Random.nextDouble(-1.5, 1.5)
            val noteZ = entity.z + entityWidth * Random.nextDouble(-1.5, 1.5)
            val noteY = entity.y + entityHeight + entityHeight * Random.nextDouble(-0.75, 0.75)

            ModPacketHandler.messageNearbyPlayers(
                SpawnNotePacket(
                    sound.location,
                    randomPitch,
                    noteX,
                    noteY,
                    noteZ
                ),
                entity.level() as ServerLevel,
                Vec3(entity.x, entity.y, entity.z),
                48.0
            )
        }

        AdvancementTriggers.hitWithInstrument(player as ServerPlayer)

        return false
    }

    override fun inventoryTick(pStack: ItemStack, pLevel: Level, pEntity: Entity, pSlotId: Int, pIsSelected: Boolean) {
        if (pEntity !is LocalPlayer) return

        if (pEntity.mainHandItem != pStack && pEntity.offhandItem != pStack) return

        val lookPitch = pEntity.lookAngle.y
        val pitch = lookPitch.toFloat().map(-1f, 1f, 0.5f, 2f)
        val note = Note.getFromPitch(pitch)

        pEntity.displayClientMessage(
            Component.literal(note.displayName).withColor(note.rgb),
            true
        )
    }

}